### PR TITLE
Adding a recipe to simplify assertj assertions with contains

### PR DESCRIPTION
**Pull request by pibizza**

### CODE REVIEW

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.J.MethodInvocation;
+
+import java.time.Duration;
+import java.util.function.Supplier;
+
+public class UseExplicitContains extends Recipe {
+    @Override
+    public String getDisplayName() {
+
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:37]:
```suggestion
```

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    @Override
+    public String getDescription() {
+        return "Convert AssertJ `assertThat(collection.contains(element)).isTrue()` with assertThat(collection).contains(element) "
+        		+ "and `assertThat(collection.contains(element)).isFalse()` with assertThat(collection).doesNotContain(element).";
+    }
+
+  @Override
+  public Duration getEstimatedEffortPerOccurrence() {
+    return Duration.ofMinutes(5);
+  }
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:50]:
```suggestion
    @Override
    public Duration getEstimatedEffortPerOccurrence() {
        return Duration.ofMinutes(5);
    }
```

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    public static class UseExplicitContainsVisitor extends JavaIsoVisitor<ExecutionContext> {
+        private static final Supplier<JavaParser> ASSERTJ_JAVA_PARSER = () -> JavaParser.fromJavaVersion().classpath("assertj-core").build();
+
+        private static final MethodMatcher ASSERT_THAT = new MethodMatcher("org.assertj.core.api.Assertions" + " assertThat(..)");
+        private static final MethodMatcher IS_TRUE = new MethodMatcher("org.assertj.core.api.AbstractBooleanAssert" + " isTrue()");
+        private static final MethodMatcher IS_FALSE = new MethodMatcher("org.assertj.core.api.AbstractBooleanAssert" + " isFalse()");
+        private static final MethodMatcher CONTAINS = new MethodMatcher("java.util.Collection contains(..)", true);
+
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:71]:
```suggestion
        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation m, ExecutionContext ctx) {
            J.MethodInvocation method = super.visitMethodInvocation(m, ctx);
```

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            	"assertThat(#{any()}).doesNotContain(#{any()});";
+            JavaTemplate builtTemplate = JavaTemplate.builder(this::getCursor, template)
+			        .javaParser(ASSERTJ_JAVA_PARSER)
+			        .build();
+			MethodInvocation withTemplate = method.withTemplate(
+            		builtTemplate,
+                    method.getCoordinates().replace(),
+                    list, 
+                    element);
+			return withTemplate;
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:107]:
```suggestion
			return method.withTemplate(
            		builtTemplate,
                    method.getCoordinates().replace(),
                    list, 
                    element);
```

File: src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import org.openrewrite.test.RewriteTest;
+
+import static org.openrewrite.java.Assertions.java;
+
+class UseExplicitContainsTest implements RewriteTest {
+
+    @Override
+    public void defaults(RecipeSpec spec) {
+        spec
+          .parser(JavaParser.fromJavaVersion().classpath("junit"))
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java:30]: You'll likely want to add AssertJ here to resolve the missing types in the tests.

File: src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	                	Collection collection = new ArrayList();
+	                	collection.add("3");
+	                    assertThat(collection).contains("3");
+	                }
+	            }
+              """
+          )
+        );
+    }
+
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java:78]: Would be good to also add a test using that should not be changed, while still being close to what the recipe matches, so something like: `assertThat(...).isTrue()` where `...` is still a method invocation, but not one to `contains`.

**Review by timtebeek — CHANGES_REQUESTED** [review]
Great start & nice addition to the suite of recipes. Some suggestions you can apply already, before you make further changes.

File: src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java (hunk 5 above)
Comment (pibizza) [reply src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java]: Done, working like a charm

File: src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java (hunk 6 above)
Comment (pibizza) [reply src/test/java/org/openrewrite/java/testing/assertj/UseExplicitContainsTest.java:78]: Done. I have added a test for isFalse/contains -> doesNotContain, and a test to show that it only affects contains/isTrue and contains/isFalse

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new UseExplicitContainsVisitor();
+    }
+
+    public static class UseExplicitContainsVisitor extends JavaIsoVisitor<ExecutionContext> {
+        private static final Supplier<JavaParser> ASSERTJ_JAVA_PARSER = () -> JavaParser.fromJavaVersion().classpath("assertj-core").build();
+
+        private static final MethodMatcher ASSERT_THAT = new MethodMatcher("org.assertj.core.api.Assertions" + " assertThat(..)");
+        private static final MethodMatcher IS_TRUE = new MethodMatcher("org.assertj.core.api.AbstractBooleanAssert" + " isTrue()");
+        private static final MethodMatcher IS_FALSE = new MethodMatcher("org.assertj.core.api.AbstractBooleanAssert" + " isFalse()");
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:66]: Might not need the string concatenation here.

Comment (pibizza) [reply src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java]: Fixed.

File: src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation m, ExecutionContext ctx) {
+            J.MethodInvocation method = super.visitMethodInvocation(m, ctx);
+        	boolean isTrue = IS_TRUE.matches(method);
+        	boolean isFalse = IS_FALSE.matches(method);
+            if (!isTrue && !isFalse) {
+                return method;
+            }
+
```
Comment (sambsnyd) [inline src/main/java/org/openrewrite/java/testing/assertj/UseExplicitContains.java:76]:
```suggestion
            if (!IS_TRUE.matches(method)&& !IS_FALSE.matches(method)) {
                return method;
            }            
```

**Comment by sambsnyd** [comment]
Thanks for your contribution @pibizza!

**Pull request merged**